### DISCUSSION TITLE
Explain volume usage and snapshots please

### DISCUSSION THREAD

**Discussion by Hr46ph** [longhorn/longhorn#10088] [Q&A] [answered]
Could someone either point me at a good explanation about how volume actual size and size limit work when snapshots exist?

To elaborate, I noticed that the snapshot size is counted towards actual size. When I delete a snapshot, the actual size goes down. Sometimes, I see a volume actual size can be larger than the volume size limit. Ive seen applications write data to that volume without a problem with the actual size growing well beyond the size limit. while another time I had an application crash because the size limit was reached.

This makes it confusing for me to determine when a volume is actually really about to grow beyond its real hard size limit (ie, app going to crash or denied write action), or to determine how much is actually really still free, even though the actual size is showing larger than the volume size limit.

**Comment by innobead** [comment]
Did you mean https://longhorn.io/docs/1.7.2/references/settings/#snapshot-maximum-count?

cc @FrankYang0529

**Reply by Hr46ph** [reply]
Thanks, but no I do not mean maximum snapshot count.

I mean for example, when I create a volume of 10 GB and there is 1GB of application data stored into it, lets say I make a snapshot every hour. The application changes 100 MB on average every hour. So after 10 hours I will have 1GB of snapshot data on top of the 1GB actual data. The volume actual size will show as 2GB even though there is only 1GB of actual application data. The other 1GB is in 10 snapshots. When I start deleting the snapshots 1 at a time, I can see the actual size coming down by 100MB with each snapshot. At least, this is how I think it works from the behavior I have seen by making and deleting snapshots.

So the question is, when I look at the volumes in Longhorn it is not obvious how much the actual data is, and how much of it is snapshot data. Given the example above, is there a way to see this somehwere?

**Reply by derekbit** [reply, marked as answer]
Currently, Longhorn actual size is the sum the sizes of volume head (active data) and snapshots. We can improve this.
https://longhorn.io/docs/1.7.2/nodes-and-volumes/volumes/volume-size

For now, you can get the statistics by https://longhorn.io/docs/1.7.2/monitoring/kubelet-volume-metrics/.

**Reply by Hr46ph** [reply]
Thank you.

I am using Prometheus and Grafana, I can see those metrics.

Don't mind me asking but just to be absolutely clear and so that I can trust this:
Snapshot sizes will never cause a volume to run out of space as long as the active data is less than the volume size limit?

And as a sidenote:
I wanted to complement you and your team for doing an amazing job on all the questions and issues being posted. Thanks to all of you!

**Reply by Hr46ph** [reply]
Something I noticed is that Longhorn often shows a much larger used size than Grafana, without snapshots on that volume.
